### PR TITLE
LIVE-4410: feat: allow one-at-a-time message processing

### PR DESCRIPTION
**Pull request by eric-gt**
use generators to interactively pull messages from subscriptions if a flag has been set in the handler metadata

[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=W5BkhfK)

### CODE REVIEW

**Review by Danglebary — APPROVED** [review]

File: lib/server/server-google-pubsub.ts [hunk 1]
```diff
@@ -286,6 +308,14 @@ export class GooglePubSubTransport extends Server implements CustomTransportStra
         ];
     };
 
+    private async handleMessageSync([pattern, iterator]: [
+        string,
+        AsyncGenerator<GooglePubSubMessage>,
+    ]) {
+        const message = await iterator.next();
+        const data = this.deserializeAndAddContext([pattern, message.value]);
+        this.handleMessage(data);
```
Comment (haroldwaters) [inline lib/server/server-google-pubsub.ts:317]: Does this also need to be awaited?

File: lib/server/server-google-pubsub.ts [hunk 2]
```diff
@@ -105,6 +111,11 @@ export class GooglePubSubTransport extends Server implements CustomTransportStra
 
     public listen(callback: () => void): void {
         this.bindHandlers(callback);
+        //for one at a time messages, pull events from their iterators and handle them
+        // synchronously
+        if (this.iterators) {
+            this.iterators.map((iterator) => this.handleMessageSync(iterator));
+        }
```
Comment (haroldwaters) [inline lib/server/server-google-pubsub.ts:118]: Tuck this into a helper method, `bindSyncHandlers` or something like that.

File: lib/server/server-google-pubsub.ts [hunk 3]
```diff
@@ -286,6 +308,14 @@ export class GooglePubSubTransport extends Server implements CustomTransportStra
         ];
     };
 
+    private async handleMessageSync([pattern, iterator]: [
+        string,
+        AsyncGenerator<GooglePubSubMessage>,
+    ]) {
+        const message = await iterator.next();
```
Comment (haroldwaters) [inline lib/server/server-google-pubsub.ts:315]: This feels odd - doesn't this need to be in a `for await...of` loop?

Comment (AndrewBird81) [reply lib/server/server-google-pubsub.ts:315]: Also, what if the `message` here is `done`?

Comment (eric-gt) [reply lib/server/server-google-pubsub.ts]: good point, I switched to `for await..of` which also automatically handles `done` iterables

Comment (eric-gt) [reply lib/server/server-google-pubsub.ts]: But this raises a good point, if our subscription's message queue ever runs dry, we're going to unsubscribe from the listener and end the loop, right?

Comment (haroldwaters) [reply lib/server/server-google-pubsub.ts]: It is, for all incense and porpoises, an infinite loop - you'll only leave if you error out or use a keyword like `return` or `break`

File: lib/server/server-google-pubsub.ts [hunk 4]
```diff
@@ -286,6 +321,21 @@ export class GooglePubSubTransport extends Server implements CustomTransportStra
...
+     * @param pattern - The subscription name
+     * @param iterator - The message iterator
+     */
+    private async handleMessageSync([pattern, iterator]: [
+        string,
+        AsyncGenerator<GooglePubSubMessage>,
+    ]) {
+        for await (const message of iterator) {
+            const data = this.deserializeAndAddContext([pattern, message]);
+            this.handleMessage(data);
```
Comment (haroldwaters) [inline lib/server/server-google-pubsub.ts:335]: this should be `await`ed

**Review by haroldwaters — CHANGES_REQUESTED** [review]

File: lib/server/server-google-pubsub.ts (hunk 4 above)
Comment (eric-gt) [reply lib/server/server-google-pubsub.ts]: nvm, I see what you mean

**Review by haroldwaters — APPROVED** [review]

**Pull request merged**